common/src/sql/dbinit.sql

```diff
@@ -707,6 +707,56 @@ INSERT INTO omicron.public.user_builtin (
 );
 
 
+/*
+ * Roles built into the system
+ *
+ * You can think of a built-in role as an opaque token to which we assign a
+ * hardcoded set of permissions.  The role that we call "project.viewer"
+ * corresponds to the "viewer" role on the "project" resource.  A user that has
+ * this role on a particular Project is granted various read-only permissions on
+ * that Project.  The specific permissions associated with the role are defined
+ * in Omicron's Polar (Oso) policy file.
+ *
+ * A built-in role like "project.viewer" has four parts:
+ *
+ * * resource type: "project"
+ * * role name: "viewer"
+ * * full name: "project.viewer"
+ * * description: "Project Viewer"
+ *
+ * Internally, we can treat the tuple (resource type, role name) as a composite
+ * primary key.  Externally, we expose this as the full name.  This is
+ * consistent with RFD 43 and other IAM systems.
+ *
+ * These fields look awfully close to the identity metadata that we use for most
+ * other tables.  But they're just different enough that we can't use most of
+ * the same abstractions:
+ *
+ * * "id": We have no need for a uuid because the (role_name, full_name) is
+ *   already unique and immutable.
+ * * "name": What we call "full name" above could instead be called "name",
+ *   which would be consistent with other identity metadata.  But it's not a
+ *   legal "name" because of the period, and it would be confusing to have
+ *   "resource type", "role name", and "name".
+ * * "time_created": not that useful because it's whenever the system was
+ *   initialized, and we have plenty of other timestamps for that
+ * * "time_modified": does not apply because the role cannot be changed
+ * * "time_deleted" does not apply because the role cannot be deleted
+ *
+ * If the set of roles and their permissions are fixed, why store them in the
+ * database at all?  Because what's dynamic is the assignment of roles to users.
+ * We [will] have a separate table that says "user U has role ROLE on resource
+ * RESOURCE".  How do we represent the ROLE part of this association?  We use a
+ * foreign key into this "role_builtin" table.
```

Comment:
Since we don't use true foreign keys, and we're not storing any mutable data in the table, I'm not sure if we get much benefit from having this table versus keeping the table in our code. The "foreign key" columns in the user permissions table can just as easily be used to reference into a map in Nexus. Are there benefits we expect from having it in the database?

Reply:
You're right that the table doesn't buy us a whole lot today and I've been on the fence about whether to even do this.  I initially decided to make this table as part of standard normalization.  In retrospect that seems dubious if the primary key comprises the entire row.  I later added metadata from RFD 43 -- the names of the roles that we'd show in the UI.  Another plus side to doing this now is that I think this schema resembles what we might do in the future with custom roles, and it'd be nice to get some experience with it.  This too feels like a weak reason.

That said, if we _are_ going to have a table of this information, even if it's immutable, I think it's simpler to keep it in the database with the rest of our tables than to have a second set of read-only tables that live inside Nexus.  That's especially true if the database tables refer to entries in the in-memory tables.

One case where I can imagine an in-database table would actually differ from an in-memory table is if we add a new built-in role.  In that case, the first Nexus that we deploy with that role will insert it into the database.  Other Nexus instances will immediately see this role, e.g., if you were to list roles.  However, their Oso Policy files won't know about it, so I don't think they'll ever look for this role.  So this improvement is pretty marginal.

In summary, I'm still on the fence about it, but it seems to me slightly better than the alternative.  What do you think?

Reply:
FWIW, having fleshed out #520, I still like that this is a separate table.  Nothing's really changed but it feels more consistent with what I expect we'll wind up doing with custom roles, and at this point it would be more work to undo it without much obvious benefit.

Reply:
Either is probably fine